internal/component/prometheus/remote/queue/filequeue/filequeue.go

```diff
@@ -0,0 +1,190 @@
+}
+
+// NewQueue returns a implementation of FileStorage.
+func NewQueue(directory string, out func(ctx context.Context, dh types.DataHandle), logger log.Logger) (types.FileStorage, error) {
+	err := os.MkdirAll(directory, 0777)
+	if err != nil {
+		return nil, err
+	}
+
+	// We dont actually support uncommitted but I think its good to at least have some naming to avoid parsing random files
+	// that get installed into the system.
+	matches, _ := filepath.Glob(filepath.Join(directory, "*.committed"))
+	ids := make([]int, len(matches))
+
+	// Try and grab the id from each file.
+	// 1.committed
+	for i, fileName := range matches {
+		id, err := strconv.Atoi(strings.ReplaceAll(filepath.Base(fileName), ".committed", ""))
+		if err != nil {
+			level.Error(logger).Log("msg", "unable to convert numeric prefix for committed file", "err", err, "file", fileName)
+			continue
+		}
+		ids[i] = id
+	}
+	sort.Ints(ids)
+	var currentMaxID int
+	if len(ids) > 0 {
+		currentMaxID = ids[len(ids)-1]
+	}
+	q := &queue{
+		directory:      directory,
+		maxID:          currentMaxID,
+		logger:         logger,
+		out:            out,
+		dataQueue:      actor.NewMailbox[types.Data](),
+		existingsFiles: make([]string, 0),
+	}
+
+	// Save the existing files in `q.existingFiles`, which will be push to `out` when actor starts.
+	for _, id := range ids {
+		name := filepath.Join(directory, fmt.Sprintf("%d.committed", id))
+		q.existingsFiles = append(q.existingsFiles, name)
+	}
+	return q, nil
+}
+
+func (q *queue) Start() {
+	// Actors and mailboxes have to be started. It makes sense to combine them into one unit since they
+	// have the same lifespan.
+	q.self = actor.Combine(actor.New(q), q.dataQueue).Build()
+	q.self.Start()
+}
+
+func (q *queue) Stop() {
+	q.self.Stop()
+}
+
+// Store will add records to the dataQueue that will add the data to the filesystem. This is an unbuffered channel.
+// Its possible in the future we would want to make it a buffer of 1, but so far it hasnt been an issue in testing.
+func (q *queue) Store(ctx context.Context, meta map[string]string, data []byte) error {
+	return q.dataQueue.Send(ctx, types.Data{
+		Meta: meta,
+		Data: data,
+	})
+}
+
+// get returns the data of the file or an error if something wrong went on.
+func get(logger log.Logger, name string) (map[string]string, []byte, error) {
+	defer deleteFile(logger, name)
+	buf, err := readFile(name)
+	if err != nil {
+		return nil, nil, err
+	}
+	r := &Record{}
+	_, err = r.UnmarshalMsg(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+	return r.Meta, r.Data, nil
+}
+
+// DoWork allows most of the queue to be single threaded with work only coming in and going out via mailboxes(channels).
+func (q *queue) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Queue up our existing items.
+	for _, name := range q.existingsFiles {
+		q.out(ctx, types.DataHandle{
+			Name: name,
+			Get: func() (map[string]string, []byte, error) {
+				return get(q.logger, name)
+			},
```

Comment:
So a call to `get` will delete the file, and thus types.DataHandle.Get() can only be called once because it has a side-effect (unexpected for a method called Get, btw)

I think that's an important behaviour that needs to be documented. But also not sure if we want this... can you share some context why we do a lazy file reading via DataHandle and delete it only after it's read?

What happens if `Get` is never called for some reason? Seems like we'd leak a file and it would be picked up again on next run?

Reply:
The assumption (that I should document) is that `Get` is called when the caller is ready to process the file, to limit the amount of lost data and limit the amount of memory in use. The `out` queue (shown in upcoming PR) has a capacity of one, so if `Get` is never called then data was never sent. This allows us to err on the side of dropping data versus sending duplicates.

Reply:
Retry/Backoff is handled in the network loop. In regards to get not being called the actual implementation of the callee processes them one at a time. If Get is not called then it should be retried on restart which would requeue it. In the main PR the lifecycles of the FileQueue and the Endpoint (callee in practice), are tied together at the component level. From the filequeue perspective the endpoint processes the file sequentially, deserializing and then adding to the network buffer, and stops pulling files when the network buffer is full.

Reply:
Pop is more accurate to whats going on so will change that and add a comment.